Add a short_description to be used by mods

Desour opened the pull request:
- I've tried to implement the changes requested here: https://github.com/minetest/minetest/pull/8966#issuecomment-533924211
- There's a new `short_description` field in item def. There's also a new `get_short_description` function for item stacks.
- This PR doesn't include the feature of seeing a different tooltip when holding shift anymore. That will, if ever, be done in a separate PR.
- This PR probably fixes parts of #8956.
- If someone thinks they can do this better than me, which is likely the cause, then please go ahead (and use the code of this PR if you wish to). Maintaining this PR / series of PRs is exhausting.

## To do

This PR is a Ready for Review.

## How to test

- Go into a minimal world.
- Hover over the mese pick and the torch pick to see their description.
- Wield the mese pick or the torch and use the `item_description` chatcommand.

paramat commented:
I'd prefer to avoid another itemdef field, can we just assume the first line of the description is automatically the short description? I seem to remember that was suggested?
I'm not really convinced this long/short description topic justifies adding a new itemdef field. However i'm no expert on this topic, waiting for rubenwardy's input.

Linuxdirk commented:
> I'd prefer to avoid another itemdef field, can we just assume the first line of the description is automatically the short description? I seem to remember that was suggested?

It should be possible to set a longer description anyways.

    My cool thing
    Some feature: yes
    Another feature: foo
    An important value: 5

When you auto-generate from this, the node description becomes completely irrelevant ("My cool thing"). Mod authors should be able to provide a better one-liner (for the example "My cool thing (yes, foo, 5)" would be useful).

Or for special tools (this change should of course not be limited to nodes only).

    My cool tool
    Left click: action 1
    Right click: action 2

... will become "My cool tool" when automatically generated from first line but could be "My cool tool (left: 1, right: 2)".

The additional value should not be mandatory but it should be possible to optionally set it to whatever the mod author likes.

rubenwardy commented:
> a simpler way for now

Adding an item property is the simpler way

paramat commented:
Oh ok =) nevermind my comment then.

rubenwardy commented:
I support this, and would like to see an update

Desour commented:
:thinking: It looks like the only thing left to do was testing.
I've rebased (luckily there were no conflicts).
In minimal, I've changed the description of the mese pick and the torch, added a short description for the mese pick and added a chatcommand for testing (see first post). @Wuzzy2 will probably like his devtest game to be merged first.

rubenwardy commented:
I've added some tests. Check to see if you agree with them

rubenwardy commented:
Should this strip color from the description?

Wuzzy2 commented:
```
        short_description = "Steel Axe",
        -- Must not contain new lines.
        -- Defaults to the first line of description.
```

Is this true?

So if I do `description = "Bla"` in the item def of `example:item` and then do `local x = minetest.registered_items["example:item"].short_description`, which value will `x` have?

* If `"Bla"`: OK
* If anything else: The documentation is incorrect.

rubenwardy commented:
That is correct, and tested via the tests I added

Desour commented:
The tests are fine, thanks for adding them! : )

sfan5 commented on builtin/game/register.lua:
```diff
@@ -118,6 +118,12 @@ function core.register_item(name, itemdef)
+	itemdef.description = itemdef.description or name
+	-- default short_description to first line of description
+	itemdef.short_description = itemdef.short_description or
+		itemdef.description:split("\n", false, 1, false)[1]
```
gsub should be faster here than constructing a temporary table

Desour replied:
I'm not experienced with `gsub`. Would the following be correct:
`itemdef.description:gsub("^([^\n]*)\n(.*)", "%1")`

sfan5 replied:
`itemdef.description:gsub("\n.*","")` works just as well

Desour replied:
Thanks, done.

sfan5 commented on src/inventory.cpp:
```diff
@@ -255,6 +255,20 @@ std::string ItemStack::getDescription(IItemDefManager *itemdef) const
 
+std::string ItemStack::getShortDescription(IItemDefManager *itemdef) const
+{
+	std::string desc = metadata.getString("short_description");
```
this doesn't seem to be documented anywhere

Desour replied:
It's documented in `lua_api.txt`:
https://github.com/minetest/minetest/blob/9cec521975f55452b07aef570b5c096c436b807e/doc/lua_api.txt#L1977

sfan5 replied:
Whoops.
